### PR TITLE
8017163: G1: Refactor remembered sets

### PR DESCRIPTION
**Pull request by tschatzl**
Hi all,

  can I have reviews for this change that significantly refactors the remembered set for more scalability.

 The current G1 remembered set implementation has been designed for use cases and Java heaps and applications from 20 years ago.

Over time many problems with performance and in particular memory usage have been observed:

* adding elements to the lowest tier data structure takes a per-remembered set global lock. Measurements have shown that the applications can wait thousands of seconds acquiring these locks. While the affected threads are in most cases refinement threads so does not directly affect the application, it can still affect the ability of G1 to meet some goals needed for keeping pause times (i.e. amount of cards from the refinement buffers to be merged into the card table and then scanned during gc).

* there is a substantial memory overhead for managing the data structures: examples are
    * using separate (hash) tables for the three different types of card containers
    * there is significant unnecessary preallocation of memory for some of the card set containers
    * Containers store redundant information

* inflexibility when reusing memory: in the current implementation the different containers use different approaches to manage memory. Most use the C heap directly, some the C heap with some internal global memory pool. This in practice makes it very difficult to implement anything other than giving back memory in the collection pause. The corresponding "Free Collection Set" pause can take a significant amount of time because of that.
Also memory reuse is limited and preallocating arenas is limited (or would have to be reimplemented multiple times), stressing the C heap allocator.

* inability to support additional use cases: over time interesting ideas (e.g. JDK-8058803) came up for improving performance of remembered set management. Mostly due to redundant information everywhere and completely different handling of various aspects in the containers it is in practice impossible to implement these.

* (partial) inability to give back memory to the OS. While some of the containers use the C heap allocator, and so in some way give back memory, these implementations and handling is different for every container.

* the existing granularity of containers are unbalanced: currently there exist three tiers: "sparse", "fine" and "full". Sparse is an array of cards ranging in the hundreds maybe, "fine" is a bitmap covering a whole region and full is a bit indicating that that region should be scanned completely during GC.

The problem is that there is nothing between "no card at all" and "sparse" and in particular the difference between the capability to hold entries of "sparse" and "fine". I.e. memory usage difference when exceeding a "sparse" array (holding 128 entries at 32M regions, taking ~256 bytes) to fine that is able to hold 65k entries using 8kB is significant.
For these reason there is even a dedicated option to stop allocating more "fine" containers and just give up and use "full" instead to avoid excessive memory usage. With extremely bad consequences in pause times.

Over time some of these issues have been fixed or in many cases band-aided, and some of these fixes and ideas were the result of working on this change (e.g. JDK-8262185, JDK-8233919, JDK-8213108).

This change is effectively a rewrite of the Java heap card based part of a region's remembered set.

This initial fully working change can be roughly described with the following properties:

* use a single `ConcurrentHashTable` for the card containers of a given region. The container in use replaced (coarsened) on the fly within the CHT node, completely lock-free. This implements JDK-6949259.

* memory for a given region's remembered set for all containers (and the CHT nodes) is backed by per container type and per remembered set arena style bump-pointer allocation buffers. In this change, in the pause, memory is given back to free lists only. The implementation gives back memory to the OS concurrently to the application. Memory is still managed using the C heap memory manager though, but abstracted away and could be replaced by manual page memory management.

* there are now four different container types and one meta-container type. These four actual containers are:
  * inline pointer: the change store a few (3-5) cards in the CHT node directly and uses no extra memory.
  * array of cards: similar to the "sparse" container, an array of cards with a configurable amount of entries. However bulk allocation of memory is now managed at a lower level so there is much less waste.
  * bitmap: similar to "fine", a bitmap spanning a (sub-)range of memory
  * full: same as full, indicating for a (sub-)range of memory that all cards are to be looked at during scan. Similar to inline pointers, this uses no extra memory.
  * howl: the Howl container subdivides a given memory range into subranges where any of the other containers describing that sub-range of the heap may be stored in. This is somewhat similar to the idea suggested in JDK-8048504.

* care has been taken to minimize container memory usage, e.g. by not adding redundant information there and in general carefully specify them. They have been designed with future enhancements in mind.

In some benchmarks (where there is significant remembered set memory usage) we are seeing memory reduction to 25% of JDK 16 levels with this change. Garbage collection times are at most as long or shorter than before; most changes affecting pause times have been extracted earlier. Individiual affected phases are generally shorter now.

Testing: tier1-8 many times, manual and automated perf testing

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issues
 * [JDK-8017163](https://bugs.openjdk.java.net/browse/JDK-8017163): G1: Refactor remembered sets
 * [JDK-8048504](https://bugs.openjdk.java.net/browse/JDK-8048504): G1: Investigate replacing the coarse and fine grained data structures in the remembered sets
 * [JDK-6949259](https://bugs.openjdk.java.net/browse/JDK-6949259): G1: Merge sparse and fine remembered set hash tables

### Reviewers
 * [Stefan Johansson](https://openjdk.java.net/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to 3dd0e7afb334f93a473ea99f62fcced7d80dacea
 * [Ivan Walulya](https://openjdk.java.net/census#iwalulya) (@walulyai - Committer) ⚠️ Review applies to 6df9cf3502fc862f2d7e6e781f4a3700972813fc

### Contributors
 * Ivan Walulya `<iwalulya@openjdk.org>`
 * Thomas Schatzl `<tschatzl@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4116/head:pull/4116` \
`$ git checkout pull/4116`

Update a local copy of the PR: \
`$ git checkout pull/4116` \
`$ git pull https://git.openjdk.java.net/jdk pull/4116/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4116`

View PR using the GUI difftool: \
`$ git pr show -t 4116`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4116.diff">https://git.openjdk.java.net/jdk/pull/4116.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1RemSet.cpp
Comment (kstefanj) [inline src/hotspot/share/gc/g1/g1RemSet.cpp:1206]: Not entierly sure where to put this comment, but this closure is used both as a heap-region closure and as a closure for container-iterations. I think this should be documented somehow. Maybe both above this closure and maybe also list what closures are used for container-iteration.

**Review by kstefanj — COMMENTED** [review]
Just a few comments to get this going.

File: src/hotspot/share/gc/g1/g1RemSet.cpp
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1RemSet.cpp:1206]: There is already a small comment that reads
```
// Visitor for remembered sets, dropping entries onto the card table.
```
at the top of the class declaration. I'll expand on that.

**Comment by tschatzl** [comment]
Fyi on the "Remove prefetching of log buffers" commit: testing on one particular machine showed that for some reason performance decreased to baseline (jdk17 levels).
I.e. when I factored out the changes in (JDK-8266821)[https://bugs.openjdk.java.net/browse/JDK-8266821] I extended the prefetching from the merge remset phase only to also include the log buffers. This showed (minimal) improvements on pause time for that phase, but overall after merging back that change seems to slow down mixed gcs on that machine.
Removing the prefetching for the log buffers improves pause times again quite a bit below jdk17/mainline levels.

I can extract this particular change again if desired.

**Comment by tschatzl** [comment]
Another note, during development of this feature there obviously were optimizations and improvements that did not make the cut. I labelled them with a [gc-g1-remset](https://bugs.openjdk.java.net/issues/?jql=labels%20%3D%20gc-g1-remset) label in the bug tracker. Feel free to add or suggest other things.

File: src/hotspot/share/gc/g1/g1ServiceThread.cpp
Comment (kstefanj) [inline src/hotspot/share/gc/g1/g1ServiceThread.cpp:29]: I think all changes to g1ServiceThread.* could be reverted now when you added the state to the new task instead. Adding `is_enqueued()` to not check `next() == NULL` is a good change, but it should be made on its own.

File: src/hotspot/share/gc/g1/heapRegionRemSet.hpp
Comment (kstefanj) [inline src/hotspot/share/gc/g1/heapRegionRemSet.hpp:126]: Pre-existing, but what do you think about moving/including the comment for this part of the calculation in the function comment.

File: src/hotspot/share/gc/g1/heapRegion.cpp
Comment (kstefanj) [inline src/hotspot/share/gc/g1/heapRegion.cpp:622]: Instead of adding this scope, can't we just create the `LogStream` here and the re-use it below?

**Review by kstefanj — COMMENTED** [review]
A few more comments. Been looking more at the code in my IDE now as well and I think it looks good. I haven't looked closely at the tests yet, but very nice that you added those. I will look at this today or tomorrow. 

A general question about the testing? Have you done any testing with `VerifyRememberedSets` turned on?

File: src/hotspot/share/gc/g1/g1ServiceThread.cpp
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1ServiceThread.cpp:29]: I agree. Note that I think that these changes fix real issues (no virtual destructor with virtual methods) too.

File: src/hotspot/share/gc/g1/heapRegionRemSet.hpp
Comment (tschatzl) [reply src/hotspot/share/gc/g1/heapRegionRemSet.hpp:126]: I think the comment should stay where it is - it is confusing otherwise, and establishing the context for this seems hard. I can remove the comment if you want though.

**Comment by tschatzl** [comment]
> A general question about the testing? Have you done any testing with `VerifyRememberedSets` turned on?

Yes we did as long as there have been changes to the code path adding remembered sets. I'll do some reruns with that to be current, just in case.
We have not seen issues for a long time with missing remembered set entries, but re-checking is always good.

File: src/hotspot/share/gc/g1/heapRegionRemSet.hpp
Comment (kstefanj) [reply src/hotspot/share/gc/g1/heapRegionRemSet.hpp:126]: Removing sounds good to me :) Or fit it on one line like this:
```
      + (sizeof(HeapRegionRemSet) - sizeof(G1CardSet)) // Avoid double counting G1CardSet
```

**Review by kstefanj — COMMENTED** [review]
Took a closer look at the new tests and overall they look good, just a couple of small comments.

File: test/hotspot/gtest/gc/g1/test_g1CardSet.cpp
Comment (tschatzl) [reply test/hotspot/gtest/gc/g1/test_g1CardSet.cpp:421]: Since that test randomly adds cards, it is very hard to calculate the expected number of cards for verification if we do not know where the coarsening exactly happens. This is very complicated to test, and other tests already test the coarsening, although not in an MT context, so I would like to not spend the time for either a brittle or useless test.

File: src/hotspot/share/gc/g1/g1CardSet.cpp
Comment (kstefanj) [inline src/hotspot/share/gc/g1/g1CardSet.cpp:146]: Unused from what I can tell.
```suggestion
```

File: src/hotspot/share/gc/g1/g1CardSet.cpp
Comment (kstefanj) [inline src/hotspot/share/gc/g1/g1CardSet.cpp:839]: Unused.
```suggestion
```

File: src/hotspot/share/gc/g1/g1CardSet.hpp
Comment (kstefanj) [inline src/hotspot/share/gc/g1/g1CardSet.hpp:138]: Same as above.
```suggestion
```

File: src/hotspot/share/gc/g1/g1CardSet.hpp
Comment (kstefanj) [inline src/hotspot/share/gc/g1/g1CardSet.hpp:145]: Unused.
```suggestion
```

File: src/hotspot/share/gc/g1/g1CardSet.hpp
Comment (kstefanj) [inline src/hotspot/share/gc/g1/g1CardSet.hpp:309]:
```suggestion
```

File: src/hotspot/share/gc/g1/heapRegionRemSet.hpp
Comment (kstefanj) [inline src/hotspot/share/gc/g1/heapRegionRemSet.hpp:85]: Never used.
```suggestion
```

File: src/hotspot/share/gc/g1/heapRegionRemSet.hpp
Comment (kstefanj) [inline src/hotspot/share/gc/g1/heapRegionRemSet.hpp:140]: Please also remove the comment.
```suggestion
```

**Review by kstefanj — APPROVED** [review]
Looked through the changes again and I think they are good. As we have all of JDK 18 to test, polish it and fix any potential problems I see no reason to not approve this now. 

I found a few unused functions, please remove them unless you have some future plans for any of them.

File: src/hotspot/share/gc/g1/g1CardSet.cpp
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1CardSet.cpp:146]: Yes, after the change to add `subtract_from`...

File: src/hotspot/share/gc/g1/g1CardSet.cpp
Comment (tschatzl) [reply src/hotspot/share/gc/g1/g1CardSet.cpp:839]: I forgot to forward to this method in `HeapRegionRemSet::occupancy_less_or_equal_than()`.

**Review by walulyai — APPROVED** [review]
Looks good!

A few minor nits.

**Pull request closed without merging**
